### PR TITLE
Vasilis/ortholearner refactor

### PR DESCRIPTION
**Pull request by vasilismsr**

### CODE REVIEW

File: econml/_ortho_learner.py [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+        - integer, to specify the number of folds.
+        - :term:`CV splitter`
+        - An iterable yielding (train, test) splits as arrays of indices.
+
+        For integer/None inputs, if the treatment is discrete
+        :class:`~sklearn.model_selection.StratifiedKFold` is used, else,
+        :class:`~sklearn.model_selection.KFold` is used
+        (with a random shuffle in either case).
+
+        Unless an iterable is used, we call `split(X,T)` to generate the splits.
```
Comment (kbattocchi) [inline econml/_ortho_learner.py]: Now split is called on the concatenation of the non-None arrays of Z, W, and X (if there are any, or a column of ones otherwise).

File: econml/_ortho_learner.py [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+        self._check_input_dims(Y, T, X, W, Z, sample_weight, sample_var)
+        nuisances, fitted_inds = self._fit_nuisances(Y, T, X, W, Z, sample_weight=sample_weight)
+        self._fit_final(self._subinds_check_none(Y, fitted_inds),
+                        self._subinds_check_none(T, fitted_inds),
+                        X=self._subinds_check_none(X, fitted_inds),
+                        W=self._subinds_check_none(W, fitted_inds),
+                        Z=self._subinds_check_none(Z, fitted_inds),
+                        nuisances=tuple([self._subinds_check_none(nuis, fitted_inds) for nuis in nuisances]),
+                        sample_weight=self._subinds_check_none(sample_weight, fitted_inds),
+                        sample_var=self._subinds_check_none(sample_var, fitted_inds))
```
Comment (kbattocchi) [inline econml/_ortho_learner.py]: I'd avoid the repetition:

    [Y, T, X, W, Z, sample_weight, sample_var, *nuisances] = [self._subinds_check_none(arr, fitted_inds) 
                                                              for arr in [Y, T, X, W, Z, sample_weight, sample_var, *nuisances]]

File: econml/_rlearner.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+
+            def fit(self, Y, T, X=None, W=None, Z=None, nuisances=None, sample_weight=None, sample_var=None):
+                Y_res, T_res = nuisances
+                self._model_final.fit(X, T_res, Y_res, sample_weight=sample_weight, sample_var=sample_var)
+                return self
+
+            def predict(self, X=None):
+                return self._model_final.predict(X)
+
+            def score(self, Y, T, X=None, W=None, Z=None, nuisances=None, sample_weight=None, sample_var=None):
```
Comment (kbattocchi) [inline econml/_rlearner.py]: sample_var is not used

File: econml/_rlearner.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+        is multidimensional, then the average of the MSEs for each dimension of Y is returned.
+    """
+
+    def __init__(self, model_y, model_t, model_final,
+                 discrete_treatment, n_splits, random_state):
+        class ModelNuisance:
+            """
+            Nuisance model fits the model_y and model_t at fit time and at predict time
+            calculates the residual Y and residual T based on the fitted models and returns
+            the residuals as two nuisance parameters.
```
Comment (kbattocchi) [inline econml/_rlearner.py]: Perhaps a philosophical question, since I think the result is the same either way, but why not predict the actual Y and T (rather than calculate the residual) and perform the residualization within ModelFinal?  Wouldn't that more closely match the common description of the nuisances in this case?

File: econml/_ortho_learner.py [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+                                                                                   sample_weight=sample_weight,
+                                                                                   sample_var=sample_var))
+
+    def const_marginal_effect(self, X=None):
+        self._check_fitted_dims(X)
+        if X is None:
+            return self._model_final.predict()
+        else:
+            return self._model_final.predict(X)
+    const_marginal_effect.__doc__ = LinearCateEstimator.const_marginal_effect.__doc__
```
Comment (kbattocchi) [inline econml/_ortho_learner.py]: Perhaps we should create a decorator for this

File: econml/tests/test_ortho_learner.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+from sklearn.linear_model import LinearRegression, LassoCV, Lasso
+from sklearn.model_selection import KFold
+import numpy as np
+import unittest
+import joblib
+
+
+class TestOrthoLearner(unittest.TestCase):
+
+    def test_crossfit(self):
```
Comment (kbattocchi) [inline econml/tests/test_ortho_learner.py]: We need more tests surrounding the behavior of `_crossfit` when keyword and non-keyword arguments are None, to verify that the behavior matches the spec.

File: notebooks/OrthoLearner.ipynb [hunk 7]
```diff
@@ -0,0 +1,1126 @@
+{
```
Comment (kbattocchi) [inline notebooks/OrthoLearner.ipynb]: This doesn't look like a user-facing notebook - if it's just something that you're using for testing then don't commit it.

File: econml/_rlearner.py [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+                Y_res_pred = np.einsum('ijk,ik->ij', effects, T_res).reshape(Y_res.shape)
+                if sample_weight is not None:
+                    return np.mean(np.average((Y_res - Y_res_pred)**2, weights=sample_weight, axis=0))
+                else:
+                    return np.mean((Y_res - Y_res_pred)**2)
+
+        super().__init__(ModelNuisance(model_y, model_t),
+                         ModelFinal(model_final), discrete_treatment, n_splits, random_state)
+
+    def fit(self, Y, T, X=None, W=None, sample_weight=None, sample_var=None, *, inference=None):
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/_rlearner.py]: done

File: econml/cate_estimator.py [hunk 9]
```diff
@@ -121,21 +122,21 @@ def effect(self, X=None, *, T0, T1):
     @abc.abstractmethod
     def marginal_effect(self, T, X=None):
         """
-        Calculate the heterogeneous marginal effect ∂τ(·,·).
+        Calculate the heterogeneous marginal effect marginal(T, X).
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/cate_estimator.py]: Done in new commit

File: econml/cate_estimator.py [hunk 10]
```diff
@@ -185,20 +186,20 @@ class LinearCateEstimator(BaseCateEstimator):
     @abc.abstractmethod
     def const_marginal_effect(self, X=None):
         """
-        Calculate the constant marginal CATE θ(·).
+        Calculate the constant marginal CATE :math:`\\theta(·)`.
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/cate_estimator.py]: I thought that Unicode greek symbols are a bit restrictive and I want to move more towards latex to make it look prettier. For future math in other places of the doc, where Unicode greek symbols wont be enough.

File: econml/dml.py [hunk 11]
```diff
@@ -332,25 +97,39 @@ def __init__(self, model, is_Y):
...
                 else:
+                    if X is None:
+                        X = np.ones((n_samples, 1))
+                    if W is None:
+                        W = np.empty((n_samples, 0))
                     return hstack([X, W])
 
             def fit(self, X, W, Target, sample_weight=None):
+                if (not self._is_Y) and discrete_treatment:
+                    Target = np.matmul(Target, np.arange(1, Target.shape[1] + 1)).flatten()
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/dml.py]: I think the convention is that: if discrete_treatment=True, then we always treat treatment as a 2d array that represents the one-hot encoding. I think we should be explicit about that.

Flattening here is good, does not affect the above convention.

File: econml/dml.py [hunk 12]
```diff
@@ -380,21 +160,14 @@ def fit(self, X, T_res, Y_res, sample_weight=None, sample_var=None):
...
-                return cross_product(F, T)
-
             def predict(self, X):
-                X, T = broadcast_unit_treatments(X, self._d_t[0] if self._d_t else 1)
-                prediction = self._model.predict(self._combine(X, T, fitting=False))
-                return reshape_treatmentwise_effects(prediction - self._intercept if self._intercept else prediction,
+                F = self._featurizer.transform(X) if X is not None else np.ones((1, 1))
+                F, T = broadcast_unit_treatments(F, self._d_t[0] if self._d_t else 1)
+                prediction = self._model.predict(cross_product(F, T))
+                prediction -= self._intercept if self._intercept is not None else 0
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/dml.py]: done

File: doc/_templates/autosummary/module.rst [hunk 13]
```diff
@@ -3,3 +3,6 @@
 
 .. automodule:: {{ fullname }}
     :members:
+    :private-members:
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply doc/_templates/autosummary/module.rst]: I expect that _OrthoLearner, _RLearner and _crosffit would be classes and methods that developers that might contribute would want to depend on. And even us, these might be classes that several of us might want to depend on in the future. So I want them to have crystal clear and visible documentation. Also in scikit-learn I like following the path of parents in the docs and code to understand what the class is doing. And this docs here are very useful for understanding the logic of the code. So I want at least these three things to be visible in the module docstring.

File: econml/_ortho_learner.py [hunk 14]
```diff
@@ -0,0 +1,584 @@
...
+        args_train = ()
+        args_test = ()
+        for var in args:
+            args_train += (var[train_idxs],) if var is not None else (None,)
+            args_test += (var[test_idxs],) if var is not None else (None,)
+
+        kwargs_train = {}
+        kwargs_test = {}
+        for key, var in kwargs.items():
+            if var is not None:
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/_ortho_learner.py]: I want the coders that use the _OrthoLearner to not have to have all variables in the signature of their models. For instance the _RLearner can ommit the Z. This wont cause any user error. If the user has passed in a Z=None, in the fit method and the method does not accept a Z, then it will throw an error. Also _crossfit is only called within the _OrthoLearner and there I am making sure that all variables except Y, T are called as keyword. This has to happen. Crossfit should only be called with all variables except Y, T as keywork variables. Also, every method that implements the nuisance and final models should have a signature were all variables except Y, T are keyword. I believe I made that explicit in the _OrthoLearner docstring. So the docstring you metnion of fit(X, Y, W) is not correct. It has to be fit(X=X, Y=Y, W=W). This is the contract. Shall I make it more explicit somewhere?

File: econml/_ortho_learner.py [hunk 15]
```diff
@@ -0,0 +1,584 @@
...
+        .. math ::
+            \\mathbb{E}[\\ell(V; \\theta(X), h(V))]
+
+        where :math:`V` are all the random variables and h is a vector of nuisance functions. Alternatively,
+        the class would also work if :math:`\\theta(X)` is the solution to a set of moment equations that
+        also depend on nuisance functions :math:`h`.
+
+    2.  To estimate :math:`\\theta(X)` we first fit the h functions can calculate :math:`h(V_i)` for each sample
+        :math:`i` in a crossfit manner:
+
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/_ortho_learner.py]: Added a more elaborate description of the crossfit training.

File: econml/_ortho_learner.py [hunk 16]
```diff
@@ -0,0 +1,584 @@
...
+    Parameters
+    ----------
+    model : object
+        An object that supports fit and predict. Fit must accept all the args
+        and the keyword arguments kwargs. Similarly predict must all accept
+        all the args as arguments and kwards as keyword arguments. The fit
+        function estimates a model of the nuisance function, based on the input
+        data to fit. Predict evaluates the fitted nuisance function on the input
+        data to predict.
+    folds : list of tuples
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/_ortho_learner.py]: I added all these extra tests and corresponding unit-tests that check that these are raised. Valid points. It does add some extra computation ot make these checks. If you think there is a faster way to check these we can improve in future folds if necessary. For now I think it's ok.

File: econml/_ortho_learner.py [hunk 17]
```diff
@@ -0,0 +1,584 @@
...
+        if not isinstance(nuisance_temp, tuple):
+            nuisance_temp = (nuisance_temp,)
+
+        if idx == 0:
+            nuisances = tuple([np.full((args[0].shape[0],) + nuis.shape[1:], np.nan) for nuis in nuisance_temp])
+
+        for it, nuis in enumerate(nuisance_temp):
+            nuisances[it][test_idxs] = nuis
+
+    return nuisances, model_list, np.sort(fitted_inds.astype(int))
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/_ortho_learner.py]: Don't recall now. I got that error once. Doesn't hurt I guess.

File: econml/_ortho_learner.py [hunk 18]
```diff
@@ -0,0 +1,584 @@
...
+
+        np.random.seed(123)
+        X = np.random.normal(size=(100, 3))
+        import scipy.special
+        from sklearn.linear_model import LogisticRegression
+        T = np.random.binomial(1, scipy.special.expit(X[:, 0]))
+        y = T + X[:, 0] + np.random.normal(0, 0.01, size=(100,))
+        est = _OrthoLearner(ModelNuisance(LogisticRegression(solver='lbfgs'), LinearRegression()),
+                            ModelFinal(), n_splits=2, discrete_treatment=True, random_state=None)
+        est.fit(y, T, W=X)
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/_ortho_learner.py]: done

File: econml/_ortho_learner.py [hunk 19]
```diff
@@ -0,0 +1,584 @@
...
+    array([12.34401722])
+    >>> est.score(y, X[:, 0], W=X[:, 1:])
+    0.00727995424098179
+    >>> est.model_final.model
+    LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
+         normalize=False)
+    >>> est.model_final.model.coef_
+    array([1.02364993])
+
+    The following example shows how to do double machine learning with discrete treatments, using
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/_ortho_learner.py]: I think given the centrality of these classes and how many of us might use them to base other classes, I would like to have an extensive documention that keeps all the contracts that this class satisfies. If someone needs to make a change to this class that breaks any of these examples, then they need to think more before they do it. So having these examples and requiring that they are valid, might be a good think. Of course this can also happen with unit tests, but I want it to be hard for someone to change the contract that I am describing in the docstring :)

File: econml/_ortho_learner.py [hunk 20]
```diff
@@ -0,0 +1,584 @@
...
+        self._discrete_treatment = discrete_treatment
+        self._random_state = check_random_state(random_state)
+        if discrete_treatment:
+            self._label_encoder = LabelEncoder()
+            self._one_hot_encoder = OneHotEncoder(categories='auto', sparse=False)
+        super().__init__()
+
+    def _check_input_dims(self, Y, T, X=None, W=None, Z=None, sample_weight=None, sample_var=None):
+        assert shape(Y)[0] == shape(T)[0], "Dimension mis-match!"
+        assert (X is None) or (X.shape[0] == Y.shape[0]), "Dimension mis-match!"
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/_ortho_learner.py]: good point.

File: econml/_ortho_learner.py [hunk 21]
```diff
@@ -0,0 +1,584 @@
...
+        inference: string, `Inference` instance, or None
+            Method for performing inference.  This estimator supports 'bootstrap'
+            (or an instance of `BootstrapInference`).
+
+        Returns
+        -------
+        self : _OrthoLearner instance
+        """
+        self._check_input_dims(Y, T, X, W, Z, sample_weight, sample_var)
+        nuisances, fitted_inds = self._fit_nuisances(Y, T, X, W, Z, sample_weight=sample_weight)
```
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/_ortho_learner.py]: sample_var is mostly used for inference purposes. So didn't think it was necessary. I say we leave it this way and if we need we can make it part of the nuisance models too. I haven't found of such a reason yet.

File: econml/cate_estimator.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kbattocchi) [reply econml/cate_estimator.py]: LaTeX is certainly more flexible; the advantage of Unicode is that it's readable in tooltips inside of a code editor, while LaTeX is only easily readable in our docs.

File: econml/_ortho_learner.py (hunk 1 above)
Comment (vasilismsr) [reply econml/_ortho_learner.py]: indeed. Fixed docstring accordingly.

File: econml/_ortho_learner.py (hunk 2 above)
Comment (vasilismsr) [reply econml/_ortho_learner.py]: The nuisances is problematic here, as we want to apply the subinds function on each entry in the nuisance tuple. Don't think the above would work. Seems fine to keep it explicity for now.

File: econml/_rlearner.py (hunk 3 above)
Comment (vasilismsr) [reply econml/_rlearner.py]: The _OrthoLearner expects this signature from the score functions of the model_final.

File: econml/_rlearner.py (hunk 4 above)
Comment (vasilismsr) [reply econml/_rlearner.py]: Yeah. seem the same. Thought this was neater, but no big deal. Residuals can also be equivalently thought as the nuisances.

File: econml/_ortho_learner.py (hunk 5 above)
Comment (vasilismsr) [reply econml/_ortho_learner.py]: Hm. I'm fine without for now :) Then I need to remember we had this decorator :)

File: econml/tests/test_ortho_learner.py (hunk 6 above)
Comment (vasilismsr) [reply econml/tests/test_ortho_learner.py]: Added some more. Let me know if enough. Please propose any other concrete cases.

File: notebooks/OrthoLearner.ipynb (hunk 7 above)
Comment (vasilismsr) [reply notebooks/OrthoLearner.ipynb]: I mostly committed it because I'm working from multiple workstations. Will remove it from the PR once all is finalized and right before merge.

File: econml/utilities.py [hunk 22]
```diff
@@ -1636,6 +1636,7 @@ def intercept_(self):
     @property
     def _param_var(self):
         """
+
```
Comment (kbattocchi) [inline econml/utilities.py]: Put a description here (it will be basically identical to the return) or else sphinx will complain.  Same for `_param_stderr`.

File: econml/cate_estimator.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (vasilismsr) [reply econml/cate_estimator.py]: I would vote for latex, as I want to refer to this theta in other parts of the docstring inside math equations. I see the drawback, but still think someone can get what this symbol is.

File: econml/utilities.py (hunk 22 above)
Comment (vasilismsr) [reply econml/utilities.py]: done

File: notebooks/OrthoLearner.ipynb (hunk 7 above)
Comment (vasilismsr) [reply notebooks/OrthoLearner.ipynb]: Removed the notebook.

File: econml/cate_estimator.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (vsyrgkanis) [reply econml/cate_estimator.py]: Nope, we could do that too. That should make it more uniform.

[the remark this answers is not in the document]
Comment (vsyrgkanis) [reply econml/cate_estimator.py]: changed in new commit.

**Comment by kbattocchi** [comment]
Please squash when merging.
